### ISSUE TITLE
Why is DALI much slower than pytorch dataloader for ImageNet loading

### ISSUE THREAD

**Issue by LuoXin-s** [NVIDIA/DALI#4834]
### Describe the question.

I have tried using DALI loading ImageNet, but I get a extremely worse results.

Below are my test code for usual pytorch dataloader.
```python
import sys
import gc
import time

import torch
from torch.utils.data.dataloader import DataLoader

import torchvision.transforms as transforms
from torchvision.datasets.imagenet import ImageNet

from timm.data.constants import IMAGENET_DEFAULT_MEAN, IMAGENET_DEFAULT_STD

# Timing utilities
start_time = None


def start_timer():
    global start_time
    gc.collect()
    torch.cuda.empty_cache()
    torch.cuda.reset_max_memory_allocated()
    torch.cuda.synchronize()
    start_time = time.time()


def end_timer_and_print(local_msg):
    torch.cuda.synchronize()
    end_time = time.time()

    print(local_msg)
    print("Total execution time = {:.3f} sec".format(end_time - start_time))
    print("Max memory used by tensors = {} bytes".format(torch.cuda.max_memory_allocated()))
    sys.stdout.flush()

transform = transforms.Compose([
            transforms.RandomResizedCrop(224, scale=(0.2, 1.0), interpolation=3),  # 3 is bicubic
            transforms.RandomHorizontalFlip(),
            transforms.ToTensor(),
            transforms.Normalize(mean=IMAGENET_DEFAULT_MEAN, std=IMAGENET_DEFAULT_STD)])

dataset = ImageNet(/path/your/ImageNet, split='train', transform=transform)

data_loader = DataLoader(dataset=dataset, batch_size=128, num_workers=8, drop_last=True)

start_timer()
for data in data_loader:
    end_timer_and_print('data loading...')
    start_timer()
```
and for DALI.
```python
from nvidia.dali.plugin.pytorch import DALIGenericIterator, LastBatchPolicy
from nvidia.dali.pipeline import pipeline_def
import nvidia.dali.fn as fn
import nvidia.dali.types as types

import gc
import time
import sys

import torch

# Timing utilities
start_time = None


def start_timer():
    global start_time
    gc.collect()
    torch.cuda.empty_cache()
    torch.cuda.reset_max_memory_allocated()
    torch.cuda.synchronize()
    start_time = time.time()


def end_timer_and_print(local_msg):
    torch.cuda.synchronize()
    end_time = time.time()

    print(local_msg)
    print("Total execution time = {:.3f} sec".format(end_time - start_time))
    print("Max memory used by tensors = {} bytes".format(torch.cuda.max_memory_allocated()))
    sys.stdout.flush()

@pipeline_def
def imagenet_dali_pipeline(data_dir, crop, size, shard_id, num_shards, dali_cpu=False, is_training=True):
    images, labels = fn.readers.file(file_root=data_dir,
                                    shard_id=shard_id,
                                    num_shards=num_shards,
                                    random_shuffle=is_training,
                                    pad_last_batch=True,
                                    name="Reader")
    dali_device = 'cpu' if dali_cpu else 'gpu'
    decoder_device = 'cpu' if dali_cpu else 'mixed'
    # ask nvJPEG to preallocate memory for the biggest sample in ImageNet for CPU and GPU to avoid reallocations in runtime
    device_memory_padding = 211025920 if decoder_device == 'mixed' else 0
    host_memory_padding = 140544512 if decoder_device == 'mixed' else 0
    # ask HW NVJPEG to allocate memory ahead for the biggest image in the data set to avoid reallocations in runtime
    preallocate_width_hint = 5980 if decoder_device == 'mixed' else 0
    preallocate_height_hint = 6430 if decoder_device == 'mixed' else 0
    if is_training:
        images = fn.decoders.image_random_crop(images,
                                            device=decoder_device, output_type=types.RGB,
                                            device_memory_padding=device_memory_padding,
                                            host_memory_padding=host_memory_padding,
                                            preallocate_width_hint=preallocate_width_hint,
                                            preallocate_height_hint=preallocate_height_hint,
                                            random_aspect_ratio=[0.8, 1.25],
                                            random_area=[0.2, 1.0],
                                            num_attempts=100)
        images = fn.resize(images,
                        device=dali_device,
                        resize_x=crop,
                        resize_y=crop,
                        interp_type=types.INTERP_TRIANGULAR)
        mirror = fn.random.coin_flip(probability=0.5)
    else:
        images = fn.decoders.image(images,
                                device=decoder_device,
                                output_type=types.RGB)
        images = fn.resize(images,
                        device=dali_device,
                        size=size,
                        mode="not_smaller",
                        interp_type=types.INTERP_TRIANGULAR)
        mirror = False

    images = fn.crop_mirror_normalize(images.gpu(),
                                    dtype=types.FLOAT,
                                    output_layout="CHW",
                                    crop=(crop, crop),
                                    mean=[0.485 * 255, 0.456 * 255, 0.406 * 255],
                                    std=[0.229 * 255, 0.224 * 255, 0.225 * 255],
                                    mirror=mirror)
    labels = labels.gpu()
    return images, labels

pipe = imagenet_dali_pipeline(batch_size=128,
                                num_threads=8,
                                device_id=0,
                                seed=0,
                                data_dir='/path/your/ImageNet',
                                crop=224,
                                size=256,
                                dali_cpu=False,
                                shard_id=0,
                                num_shards=1,
                                is_training=True)
pipe.build()
data_loader = DALIGenericIterator(pipe, ["img", "label"], reader_name="Reader", last_batch_policy=LastBatchPolicy.DROP, auto_reset=True)

start_timer()
for data in data_loader:
    end_timer_and_print('data loading...')
    start_timer()
```

DALI one takes about 0.2 second to load a batch, and it is 0.08 second to load a batch for naive dataloader.

below are my environments:
```
Collecting environment information...
PyTorch version: 2.0.0+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.35

Python version: 3.10.11 (main, Apr 20 2023, 19:02:41) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-60-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.0.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-80GB
Nvidia driver version: 525.60.13
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 57 bits virtual
Byte Order:                      Little Endian
CPU(s):                          64
On-line CPU(s) list:             0-63
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
CPU family:                      6
Model:                           106
Thread(s) per core:              1
Core(s) per socket:              32
Socket(s):                       2
Stepping:                        6
CPU max MHz:                     3400.0000
CPU min MHz:                     800.0000
BogoMIPS:                        5200.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
L1d cache:                       3 MiB (64 instances)
L1i cache:                       2 MiB (64 instances)
L2 cache:                        80 MiB (64 instances)
L3 cache:                        96 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-31
NUMA node1 CPU(s):               32-63
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT disabled
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
```

I have DALI with cuda 12.

Does I have any wrong configuration?

### Check for duplicates

- [X] I have searched the [open bugs/issues](https://github.com/NVIDIA/DALI/issues) and have found no duplicates for this bug report

**Comment by mzient** [comment]
Hello @LuoXin-s 
The results are quite unexpected.

First of all, try the following parameters for DALI's resize:
`interp_type=dali.types.INTERP_CUBIC, antialias=False`.

Also, the amount of additional synchronization in your start_timer and end_timer_and_print routines is substantial and may affect the execution time. I'd recommend running a few warm-up iterations, start measurement and run a larger number of iterations - perhaps a few hundreds - and get the average batch time from there. Could you check if it helps on your system?

**Comment by LuoXin-s** [comment]
Hi @JanuszL ,

Thank you for your tests and advice. My program is running at a GPU cluster, so datasets are accessed through network. When I load datasets into /tmp of computing node (so it is local storage access now) before testing, the speedup of DALI is obvious, I see aroud 6x speed up in data loading process.

So DALI is actually higher demanding in reading data from 'disc'?

**Comment by JanuszL** [comment]
Hi @LuoXin-s,

In some cases, DALI may not be able to hide the latency of access to the files on the network drive (as it uses only one thread to perform the read operation while the PyTorch dataloader may read as many files in parallel as it has threads). On top of that it mmaps files to avoid unnecessary memcopies. You can set `dont_use_mmap=True` and see if that helps.

**Comment by LuoXin-s** [comment]
@JanuszL Yeah, thanks for your explanation. Currently, I solved slow loading problem by saving dataset with webdataset, i'll try `dont_use_mmap=True ` in later.

**Comment by noamsgl** [comment]
With WebDataset there is no GPU-Direct-Storage support, so what is the benefit of using DALI in the case of WebDataset?

**Comment by JanuszL** [comment]
@noamsgl - the main advantage of WebDataset is the more disc-friendly access patterns. Instead of handling multiple small files, you open one (or just a handful) of bigger ones, and a sequential read of these is much faster than random access to separate files. On top of that, samples are just concatenated inside the tar archive, so there is no additional parsing/deserialization step needed to access the raw data.
